library/psa_crypto_random.h

```diff
@@ -0,0 +1,218 @@
+ */
+/*
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#ifndef PSA_CRYPTO_RANDOM_H
+#define PSA_CRYPTO_RANDOM_H
+
+#if defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
+
+#include <string.h>
+#include <mbedtls/entropy.h> // only for error codes
+#include <psa/crypto.h>
+
+typedef mbedtls_psa_external_random_context_t mbedtls_psa_random_context_t;
+
+static inline int mbedtls_psa_get_random( void *p_rng,
+                                          unsigned char *output,
+                                          size_t output_size )
+{
+    (void) p_rng;
+    psa_status_t status = psa_generate_random( output, output_size );
+    if( status == PSA_SUCCESS )
+        return( 0 );
+    else
+        return( MBEDTLS_ERR_ENTROPY_SOURCE_FAILED );
+}
+
+static inline void *mbedtls_psa_random_state( mbedtls_psa_random_context_t *rng )
+{
+    (void) rng;
+    return( NULL );
+}
+
+#else /* MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG */
+
+/* Choose a DRBG based on configuration and availability */
+#if defined(MBEDTLS_PSA_HMAC_DRBG_MD_TYPE)
+
+#include "mbedtls/hmac_drbg.h"
+
+#elif defined(MBEDTLS_CTR_DRBG_C)
+
+#include "mbedtls/ctr_drbg.h"
+
+#elif defined(MBEDTLS_HMAC_DRBG_C)
+
+#include "mbedtls/hmac_drbg.h"
+#if defined(MBEDTLS_SHA512_C) && defined(MBEDTLS_SHA256_C)
+#include <limits.h>
+#if SIZE_MAX > 0xffffffff
+/* Looks like a 64-bit system, so prefer SHA-512. */
+#define MBEDTLS_PSA_HMAC_DRBG_MD_TYPE MBEDTLS_MD_SHA512
+#else
+/* Looks like a 32-bit system, so prefer SHA-256. */
+#define MBEDTLS_PSA_HMAC_DRBG_MD_TYPE MBEDTLS_MD_SHA256
+#endif
+#elif defined(MBEDTLS_SHA512_C)
+#define MBEDTLS_PSA_HMAC_DRBG_MD_TYPE MBEDTLS_MD_SHA512
+#elif defined(MBEDTLS_SHA256_C)
+#define MBEDTLS_PSA_HMAC_DRBG_MD_TYPE MBEDTLS_MD_SHA256
+#else
+#error "No hash algorithm available for HMAC_DBRG."
+#endif
+
+#else
+#error "No DRBG module available for the psa_crypto module."
+#endif
+
+#include "mbedtls/entropy.h"
+
+/** The type of the PSA DRBG context.
+ */
+#if defined(MBEDTLS_CTR_DRBG_C)
+typedef mbedtls_ctr_drbg_context mbedtls_psa_drbg_context_t;
+#elif defined(MBEDTLS_HMAC_DRBG_C)
+typedef mbedtls_hmac_drbg_context mbedtls_psa_drbg_context_t;
+#endif
+
+/** Initialize the PSA DRBG.
+ *
+ * \param p_rng        Pointer to the Mbed TLS DRBG state.
+ */
+static inline void mbedtls_psa_drbg_init( mbedtls_psa_drbg_context_t *p_rng )
```

Comment:
This function seems to be called only by `mbedtls_psa_random_init` thus why expose it and not just add the below code to `mbedtls_psa_random_init`?

Reply:
Because I think the code structure is clearer that way: one part of the code cares whether there's a DRBG or not, and one part cares which DRBG it is. I don't think it makes much difference, so I'm willing to change it, but I think it's minor enough to keep what's currently there unless there's a reason to change.

Reply:
In terms of code organization I agree that this is good. It is rather the fact that this is declared as an API that bothers me actually. Why is this not just in psa_crypto.c ?

Reply:
I think the code structure is good enough, but you don't. Unfortunately I don't understand what you would prefer. To move everything into `psa_crypto.c`? To move some thing into `psa_crypto.c`? To create a `psa_crypto_random.c`, but what would it contain?

Reply:
To me this thread comment is just about mbedtls_psa_drbg_init() and mbedtls_psa_drbg_free() that I would prefer to be in psa_crypto.c (let's forget about a psa_crypto_random.c, I've resolved the comment related to that) .  My reasoning is the following:  If in a Mbed TLS C module/file, I want to call `mbedtls_psa_get_random()` then I will include psa_crypto_random.h. This means that I will be able to call `mbedtls_psa_drbg_init()/_free()` and I don't see why we would allow that.

Reply:
I renamed `psa_crypto_random.h` to `psa_crypto_random_impl.h`, to clarify that this is about implementing the PSA crypto RNG, and not about using it.

I've left `mbedtls_psa_get_random` declared here for now. It should move to a public API in `include/mbedtls/*.h`, but I'll do that in step 3, which is where tests and sample programs will use it.

Reply:
> I've left `mbedtls_psa_get_random` declared here for now. 

Looks like I'll have to move the declarations of `mbedtls_psa_get_random` and `MBEDTLS_PSA_RANDOM_STATE` to `include/mbedtls/*.h` or rename them to not have an `mbedtls`/`MBEDTLS` prefix to pacify `check-names.sh` :sob:

Reply:
No, the problem is not where the declarations are located. It's that `check-names` doesn't grok
```
static int ( *const mbedtls_psa_get_random )(
    void *p_rng, unsigned char *output, size_t output_size ) =
    mbedtls_hmac_drbg_random;
```